core/arch/arm/kernel/ree_fs_ta_buffered.c

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, Linaro Limited
+ */
+/*
+ * This is a wrapper around the "REE FS" TA store.
+ * The whole TA/library is read into a temporary buffer during .open(). This
+ * allows the binary to be authenticated before any data is read and processed
+ * by the upper layer (ELF loader).
+ */
+#include <kernel/user_ta.h>
+#include <mm/mobj.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+
+#include "elf_load.h"
+#include "ree_fs_ta.h"
+
+struct user_ta_store_handle {
+	struct user_ta_store_handle *h; /* Note: a REE FS TA store handle */
+	size_t ta_size;
+	struct mobj *ta_mobj;
+	size_t offs;
+};
+
+static TEE_Result ta_open(const TEE_UUID *uuid,
+			  struct user_ta_store_handle **h)
+{
+	struct user_ta_store_handle *handle = NULL;
+	void *buf = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
+
+	handle = calloc(1, sizeof(*handle));
+	if (!handle)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = ree_fs_ta_open(uuid, &handle->h);
+	if (res)
+		goto err2;
+	res = ree_fs_ta_get_size(handle->h, &handle->ta_size);
+	if (res)
+		goto err;
+	handle->ta_mobj = tee_ta_alloc_ta_mem(handle->ta_size);
```

Comment:
I don't think this works with pager enabled.

Reply:
Why not? Memory limitation?

`CFG_REE_FS_TA_BUFFERED` is not enabled by default if pager is enabled, but it is not forbidden to enable it anyway. If it's just a memory limitation, I think it's OK, but if there is a more fundamental reason for not having `CFG_REE_FS_TA_BUFFERED=y` when `CFG_WITH_PAGER=y` then I can add:

```
ifeq ($CFG_WITH_PAGER),y)
$(call force,CFG_REE_FS_TA_BUFFERED,n)
endif
```

Reply:
The allocated memory isn't mapped here, just allocated. Compare with what happens in
https://github.com/OP-TEE/optee_os/blob/53c1131c3dee546d6d618a0f7f20586598ca032c/core/arch/arm/kernel/user_ta.c#L695
Before the memory is accessed it's first mapped with `vm_map()`

Reply:
OK but I don't need a user mapping for this memory, just a kernel one. I have no `user_ta_ctx` at this point. Well, I could have it if needed, but isn't there another way to map only for kernel?

Reply:
Not currently.

Another option is not to buffer the entire TA. Instead you save hashes page by page while calculating the overall hash. If it's OK then the subhashes are OK. The TA storage interface can afford to buffer 4kb in secure memory at a time. This approach wouldn't consume much memory and would work the same with and without pager.

Reply:
Otherwise, I think that the ELFs should be parsed in user mode for a more secure solution.